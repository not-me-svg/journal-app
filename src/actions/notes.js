import Swal from 'sweetalert2'
import { addDoc, db, collection, deleteDoc, doc, updateDoc } from "../firebase/firebase-config";
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

export const startNewNote = () => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;
        
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        try {
            const docRef = await addDoc(collection(db, `${ uid }`, "journal/notes"), newNote);

            dispatch(activateNote(docRef.id, newNote));
            dispatch(addNewNote(docRef.id, newNote));
        } catch (error) {
            console.log(error);
        }
    }
}

export const activateNote = (id, note) => {
    return {
        type: types.notesActive,
        payload: {
            id,
            ...note
        }
    }
}

export const addNewNote = (id, note) => {
    return {
        type: types.notesAddNew,
        payload: {
            id, 
            ...note
        }
    }
}

export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes(uid);

        dispatch(setNotes(notes))
    }
}

export const setNotes = (notes) => {
    return {
        type: types.notesLoad,
        payload: notes
    }
}

export const startSaveNote = (note) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;

        if (!note.url) {
            delete note.url
        }

        const noteToFirestore = { ...note }
        delete noteToFirestore.id;

        const noteRef = doc(db, `${uid}/journal/notes/${note.id}`);
        await updateDoc(noteRef, noteToFirestore);

        dispatch(refreshNote(note.id, noteToFirestore));
        Swal.fire('Saved', note.title, 'success');
    }
}

export const refreshNote = (id, note) => {
    return {
        type: types.notesUpdated,
        payload: {
            id,
            note: {
                id,
                ...note
            }
        }
    }
}

export const startUploading = (file) => {
    return async (dispatch, getState) => {
        const { active:activeNote } = getState().notes;
        
        Swal.fire({
            title: "Uploading...",
            text: "Please wait...",
            showConfirmButton: false,
            allowOutsideClick: false,
            willOpen: () => {
                Swal.showLoading();
            },
        });

        const fileUrl = await fileUpload(file);
        activeNote.url = fileUrl;

        dispatch(startSaveNote(activeNote));

        Swal.close();
        
    }
}

export const startDeleteNote = (id) => {
    return async (dispatch, getState) => {
        const uid = getState().auth.uid;

        const noteRef = doc(db,`${uid}/journal/notes/${id}`);
        await deleteDoc(noteRef);

        dispatch( deleteNote(id));
    }
}

export const deleteNote = (id) => {
    return {
        type: types.notesDelete,
        payload: id
    }
}

export const noteLogout = () => {
    return {
        type: types.notesLogOutCleaning
    }
}