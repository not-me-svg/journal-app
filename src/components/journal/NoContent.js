import React from 'react';
import { Backpack, Planet } from 'react-kawaii';

export const NoContent = () => {
  return (
        <div className="w-100 h-100 p-relative">
            <div className="ja__no-content__content d-flex flex-column ai-center">
                <h3>No content!</h3>
                <br />
                <div>
                    <Planet size={300} mood="shocked" color="#faa3aa" />
                </div>
                <br />
                <p>Select an entry or create a new one</p>
            </div>
        </div>
  );
};
