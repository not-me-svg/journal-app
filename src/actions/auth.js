import Swal from 'sweetalert2'
import { types } from "../types/types";
import { startLoading, finishLoading } from "./ui";
import { noteLogout } from './notes';
import { 
    auth,
    createUserWithEmailAndPassword,
    googleAuthProvider,
    signInWithEmailAndPassword,
    signInWithPopup,
    signOut,
    updateProfile
} from "../firebase/firebase-config";

export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        dispatch( startLoading() );

        signInWithEmailAndPassword(auth, email, password)
            .then(({ user }) => {
                dispatch( login(user.uid, user.displayName) );

                dispatch( finishLoading() );
            }).catch(err => {
                console.log(err)
                dispatch( finishLoading() );
                Swal.fire('Error', err.message, 'error');
            })
    }
}

export const startGoogleLogin = () => {
    return (dispatch) => {
        signInWithPopup(auth, googleAuthProvider)
            .then(({ user }) => {
                dispatch( login(user.uid, user.displayName) )
            }).catch(err => {
                console.log(err)
            })
    }
}

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        createUserWithEmailAndPassword(auth, email, password)
            .then(async ({ user }) => {
                await updateProfile(user, { displayName: name })

                dispatch( login(user.uid, user.displayName) )
            }).catch(err => {
                console.log(err)
                Swal.fire('Error', err.message, 'error');
            })
    }
}

export const login = (uid, displayName) => {
    return {
        type: types.login,
        payload: {
            uid,
            displayName
        }
    }
}

export const startLogOut = () => {
    return async (dispatch) => {
        await signOut(auth);

        dispatch(logout());
        dispatch(noteLogout());
    }
}

export const logout = () => ({
    type: types.logout
})