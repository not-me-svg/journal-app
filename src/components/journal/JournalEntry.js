import React from 'react';
import moment from "moment";
import { useDispatch } from "react-redux";
import { activateNote } from '../../actions/notes';

export const JournalEntry = ({ id, date, title, body, url }) => {

    const noteDate = moment(date);
    const dispatch = useDispatch();

    const handleEntryClick = () => {
        dispatch( 
            activateNote( id, {
                date, title, body, url
            })
        );
    }

    return (
        <div 
            className="ja__journal-entry c-pointer"
            onClick={ handleEntryClick }>
                {
                    url &&
                    (
                        <figure className="h-100">
                            <img 
                                src={ url.length ? url : './../assets/no-image.jpg' }
                                alt={ title }
                                className="w-100 h-100"/>
                        </figure>
                    )
                }

                <div className="ja__journal-entry__body d-flex">
                    <div className="ja__journal-entry__body__content flex-stretch">
                        <h4 className="ja__journal-entry__body__title text--sm">
                            <b>{ title.length ? title : 'No title' }</b>
                        </h4>

                        <p className="ja__journal-entry__body__text text--sm">
                            { body.length ? body : 'No content' }
                        </p>
                    </div>

                    <div className="ja__journal-entry__body__date d-flex flex-column jc-center">
                        <p className="text--xs">{ noteDate.format('dddd') }</p>

                        <p className="text--xs"><b>{ noteDate.format('Do') }</b></p>
                    </div>
                </div>
        </div>
    );
};
