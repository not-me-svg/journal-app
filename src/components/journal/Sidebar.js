import React, { useState } from 'react';
import { useDispatch, useSelector } from "react-redux";

import { JournalEntries } from './JournalEntries';
import { startLogOut } from '../../actions/auth';
import { startNewNote } from '../../actions/notes';

export const Sidebar = () => {

  const dispatch = useDispatch();
  const { name } = useSelector( state => state.auth );
  const [sideBarActive, setSideBarActive] = useState('true');

  const handleLogOut = () => {
    dispatch(startLogOut())
  }

  const handleAddNew = () => {
    dispatch(startNewNote())
  }

  const handleSidebarToggle = () => {
    setSideBarActive(!sideBarActive);
  }

  return (
    <aside className={ sideBarActive ? "ja__sidebar d-flex flex-column ai-center h-100vh" : "ja__sidebar ja__sidebar--hidden d-flex flex-column ai-center h-100vh"}>
        
        <nav className="ja__sidebar-navbar d-flex ai-center jc-space-between w-100">
            <div className="d-flex flex-stretch ai-center">
                <i className="far fa-moon"></i>
                <p className="text--sm"><b>{ name }</b></p>
            </div>

            <button 
              className="btn btn--secondary"
              onClick={ handleLogOut }>Logout</button>
        </nav>

        <div 
          className="ja__sidenav-new-entry d-flex flex-column ai-center jc-center w-100 c-pointer"
          onClick={ handleAddNew }>
            <i className="far fa-calendar-plus fa-5x"></i>

            <p>New entry</p>
        </div>

        <div className="ja__sidenav-entries flex-stretch w-100">
          <JournalEntries />
        </div>

        <button 
              className="ja__sidenav-toggle-btn btn btn--secondary btn--lg"
              onClick={ handleSidebarToggle }>
          <i className="fas fa-bars"></i>      
        </button>
    </aside>
  );
};
