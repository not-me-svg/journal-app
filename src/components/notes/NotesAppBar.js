import React from 'react';
import { useSelector, useDispatch } from "react-redux";
import { startDeleteNote, startSaveNote } from '../../actions/notes';

export const NotesAppBar = () => {

  const dispatch = useDispatch();
  const { active } = useSelector( state => state.notes );

  const handleSave = () => {
    dispatch(startSaveNote(active));
  }

  const handleDelete = () => {
    dispatch(startDeleteNote(active.id));
  }

  return (
    <div className="ja__notes-bar w-100 d-flex jc-space-between ai-center">
        <span>28 de agosto de 2020</span>

        <div>
            <button
              className="btn btn--secondary mr--sm" 
              onClick={ handleSave }>Save</button>

            <button
              className="btn btn--secondary" 
              onClick={ handleDelete }>Delete</button>
        </div>
    </div>
  );
};
