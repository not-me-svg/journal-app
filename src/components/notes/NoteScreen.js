import React, { useRef, useEffect } from 'react';
import { useSelector, useDispatch } from "react-redux";

import { NotesAppBar } from './NotesAppBar';
import { useForm } from '../../hooks/useForm';
import { activateNote, startUploading } from '../../actions/notes';

export const NoteScreen = () => {

  const dispatch = useDispatch();
  const { active:note } = useSelector( state => state.notes );
  const [ formValues, handleInputChange, reset ] = useForm(note);
  const { body, title, id } = formValues;
  
  const activeId = useRef( note.id );
 
  useEffect(() => {
    if (note.id !== activeId.current) {
      reset(note);

      activeId.current = note.id;
    }
  }, [note, reset]);
  
  useEffect(() => {
    dispatch(activateNote(formValues.id, { ...formValues }));
  }, [formValues, dispatch]);

  const handlePictureClick = (e) => {
    e.preventDefault();
    document.querySelector('#file-selector').click()
  }

  const handleFileChange = (e) => {
    const file = e.target.files[0];

    if (file) {
      dispatch(startUploading(file));
    }
  }

  return (
    <div>
      <NotesAppBar { ...note }/>

      <div className="ja__note-screen">
        <form>
          <div class="d-flex flex-column">
            <input 
              className="input"
              type="text"
              placeholder="Note title"
              autoComplete="off"
              name="title"
              value={ title }
              onChange={ handleInputChange } />

            <br />

            <textarea 
              className="textarea"
              placeholder="xxxx" 
              name="Note body" 
              id="" 
              cols="30" 
              rows="10"
              name="body"
              value={ body }
              onChange={ handleInputChange } ></textarea>
              <br />

            {
              (note.url)
              ?
              (
                <figure>
                  <img 
                      src={ note.url }
                      alt="XXX"
                      className="w-100 h-100"/>
                </figure>
              )
              :
              (
                <div>
                  <button 
                    className="btn btn--secondary mr--sm" 
                    onClick={ handlePictureClick }>Upload Picture</button>
                </div>
              )
            }
          </div>
        </form>
      </div>

      <input 
        type="file" 
        onChange={ handleFileChange }
        id="file-selector"
        name="file"
        className="hide" />
    </div>
  );
};
