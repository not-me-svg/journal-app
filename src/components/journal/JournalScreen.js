import React from 'react';
import { Sidebar } from './Sidebar';
import { NoteScreen } from '../notes/NoteScreen';
import { NoContent } from './NoContent';
import { useSelector } from 'react-redux';


export const JournalScreen = () => {

  const { active } = useSelector( state => state.notes );

  return (
    <div className="ja__main-content h-100vh">
      <Sidebar />

      <main>
        {
          ( active )
          ? ( <NoteScreen /> )
          : ( <NoContent /> )
        }
      </main>
    </div>
  );
};
