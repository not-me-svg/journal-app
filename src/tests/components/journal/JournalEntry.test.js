import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux'

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import '@testing-library/jest-dom';
import { JournalEntry } from '../../../components/journal/JournalEntry';
import { activateNote } from '../../../actions/notes';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};

let store = mockStore(initState);
store.dispatch = jest.fn();

const nota = {
    id: 10,
    date: 0,
    title: 'Hello',
    body: 'Friend',
    url: 'https://somewhere.com/foto.jpg'
};

const wrapper = mount( 
    <Provider store={ store }>
        <JournalEntry { ...nota }  /> 
    </Provider>
)

describe('Test in <JournalEntry />', () => {
    test('Should show', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('Should activate note', () => {
        wrapper.find('.ja__journal-entry').prop('onClick')();
        expect( store.dispatch ).toHaveBeenCalledWith(
            activateNote( nota.id, { ...nota } )
        );
    })
})
