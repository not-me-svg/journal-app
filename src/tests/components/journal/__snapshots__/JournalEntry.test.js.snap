// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <JournalEntry /> debe de mostrarse correctamente 1`] = `ReactWrapper {}`;

exports[`Test in <JournalEntry /> Should show 1`] = `ReactWrapper {}`;
